[primTorch] Implement conj and conj_physical

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80737
* #80736
* #80571
* #80570
* #80182
* __->__ #80358

This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

peterbell10 committed:
[primTorch] Implement conj and conj_physical

This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Update on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Update on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Update on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

mruberry commented on torch/_prims/utils.py:
```diff
@@ -55,6 +55,7 @@ def getnvFuserDtype(dtype: Union[torch.dtype, NumberTypeType]):
     torch.Tensor.numel,
     torch.Tensor.stride,
     torch.Tensor.dtype.__get__,  # type: ignore[attr-defined]
+    torch.Tensor.is_sparse.__get__,  # type: ignore[attr-defined]
```
Nice update

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1965,6 +1974,14 @@ def column_stack(tensors: TensorSequenceType) -> TensorLikeType:
+    if not input.dtype.is_complex:
+        return input
+    if input.is_sparse:
+        return prims.conj_physical(input)
```
Interesting caveat

mruberry reviewed (approved):
Cool!

fyi @SherlockNoMad and @davidberard98 to compare with https://github.com/pytorch/pytorch/pull/81014/files.

Adding `conj_physical` seems totally fine for now, but, as mentioned in the inline comments, I expect that in the future we'll just have `prims.conj` and not worry about modeling the conjugate bit.

peterbell10 committed:
Rebase on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Fix merge conflicts on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Add regsiter_decomposition statements and add _refs.fft to the tests on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Fix decomp tests on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Update on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]

peterbell10 committed:
Update on "[primTorch] Implement conj and conj_physical"


This adds `prims.conj` and `prims.conj_physical` which only accept
complex tensors, as well as `refs.conj` and `refs.conj_physical` which
pass-through non-complex values and call the appropriate `prims` for
complex types.

[ghstack-poisoned]
